roundabout/utils.js

```diff
@@ -3,7 +3,74 @@ export const MAX_EXPIRES_IN = 3 * 24 * 60 * 60 // 7 days in seconds
 export const MIN_EXPIRES_IN = 1
 export const DEFAULT_EXPIRES_IN = 3 * 24 * 60 * 60 // 3 days in seconds by default
 
-export const VALID_BUCKETS = ['dagcargo']
+export const VALID_BUCKETS_BY_KEY = ['dagcargo']
+export const VALID_R2_BUCKETS_DEFAULT = ['carpark-prod-0', 'carpark-prod-1', 'dagcargo']
+export const VALID_S3_BUCKETS_DEFAULT = ['carpark-prod-0']
+export const CF_R2_DOMAIN = 'r2.cloudflarestorage.com'
+export const AWS_S3_DOMAIN = 's3.amazonaws.com'
+
+/**
+ * Filters location claims to get the R2 buckets valid for redirect.
+ * In case of not existing any R2 bucket, verifies if there is an AWS bucket name that could be attempted in CF.
+ * 
+ * @param {Set<string>} locations 
+ * @param {object} [options]
+ * @param {string[]} [options.validR2Buckets]
+ * @param {string[]} [options.validS3Buckets]
+ */
+export function getBucketKeyPairToRedirect (locations, options = {}) {
+  const validR2Buckets = options.validR2Buckets || VALID_R2_BUCKETS_DEFAULT
+  const validS3Buckets = options.validS3Buckets || VALID_S3_BUCKETS_DEFAULT
+  // Filter by Cloudflare R2 URLs
+  const r2Urls = Array.from(locations)
+    .filter(
+      // CF Domain
+      l => l.includes(CF_R2_DOMAIN) &&
+      // Bucket name valid for CF
+      validR2Buckets.filter(b => l.includes(b)).length
+    )
+  
+  // Transform R2 URLs if existent
+  if (r2Urls.length) {
+    return r2Urls.map(url => {
+      // Format https://account-id.r2.cloudflarestorage.com/bucket-name/key
+      const domainSplit = url.split(CF_R2_DOMAIN)[1]
+      const bucketName = domainSplit.split('/')[1]
+      const key = domainSplit.split(`${bucketName}/`)[1]
+
+      return {
+        bucketName,
+        key
+      }
+    })
+  }
+
+  // Attempt S3 URL to pick bucket to try in R2
+  const s3Urls = Array.from(locations)
+    .filter(
+      // S3 Domain
+      l => l.includes(AWS_S3_DOMAIN) &&
+      // Bucket name valid for R2 attempt
+      validS3Buckets.filter(b => l.includes(b)).length
+    )
+  
+  // Transform S3 URLs if existent
+  if (s3Urls.length) {
+    return s3Urls.map(url => {
+      // Format 'https://bucket-name.s3.amazonaws.com/key'
+      const domainParts = url.split(`.${AWS_S3_DOMAIN}`)
+      const bucketName = domainParts[0].replace('https://', '')
+      const key = domainParts[1].slice(1)
```

Comment:
same here, let's use URLs and be explicit.... `3.amazonaws.com` would be a valid s3 key name i think. Even if not, a user could put anything in a location claim url

Reply:
I mean it would work as a key, but not as a CID later on? so for this to work would never be a valid key